evm/p2p/discovery.py

```diff
@@ -0,0 +1,320 @@
+            # packets. This might not be the case for a later version, but just popping
+            # the last element is good enough for now.
+            expiration = rlp.sedes.big_endian_int.deserialize(payload.pop())
+            if time.time() > expiration:
+                raise PacketExpired()
+        except DefectiveMessage as e:
+            self.logger.error('error unpacking message: {}'.format(e))
+            return
+
+        cmd = CMD_ID_MAP[cmd_id]
+        if len(payload) != cmd.elem_count - 1:
+            self.logger.error('invalid {} payload: {}'.format(cmd.name, payload))
+            return
+        node = kademlia.Node(remote_pubkey, address)
+        handler = self._get_handler(cmd)
+        handler(node, payload, mdc)
+
+    def recv_pong(self, node, payload, mdc):
+        echoed = payload[1]
+        self.kademlia.recv_pong(node, echoed)
+
+    def recv_neighbours(self, node, payload, mdc):
+        self.kademlia.recv_neighbours(node, _extract_nodes_from_payload(payload[0]))
+
+    def recv_ping(self, node, payload, mdc):
+        self.kademlia.recv_ping(node, mdc)
+
+    def recv_find_node(self, node, payload, mdc):
+        self.logger.debug('<<< find_node from {}'.format(node))
+        target = big_endian_to_int(payload[0])
+        self.kademlia.recv_find_node(node, target)
+
+    def send_ping(self, node):
+        self.logger.debug('>>> pinging {}'.format(node))
+        version = rlp.sedes.big_endian_int.serialize(PROTO_VERSION)
+        payload = [version, self.address.to_endpoint(), node.address.to_endpoint()]
+        message = _pack(CMD_PING.id, payload, self.privkey)
+        self.send(node, message)
+        return message[:32]  # return the MDC to identify pongs
+
+    def send_find_node(self, node, target_node_id):
+        target_node_id = int_to_big_endian(
+            target_node_id).rjust(kademlia.k_pubkey_size // 8, b'\0')
+        self.logger.debug('>>> find_node to {}'.format(node))
+        message = _pack(CMD_FIND_NODE.id, [target_node_id], self.privkey)
+        self.send(node, message)
+
+    def send_pong(self, node, token):
+        self.logger.debug('>>> ponging {}'.format(node))
+        payload = [node.address.to_endpoint(), token]
+        message = _pack(CMD_PONG.id, payload, self.privkey)
+        self.send(node, message)
+
+    def send_neighbours(self, node, neighbours):
+        nodes = []
+        neighbours = sorted(neighbours)
+        for n in neighbours:
+            l = n.address.to_endpoint() + [n.pubkey]
+            nodes.append(l)
+
+        max_neighbours = self._get_max_neighbours_per_packet()
+        if len(nodes) <= max_neighbours:
+            message = _pack(CMD_NEIGHBOURS.id, [nodes], self.privkey)
+            self.logger.debug('>>> neighbours to {}: {}'.format(node, neighbours))
+            self.send(node, message)
+        else:
+            for i in range(0, len(nodes), max_neighbours):
```

Comment:
The [`toolz.partition`](http://toolz.readthedocs.io/en/latest/api.html#toolz.itertoolz.partition) function will do this for you.  (make sure you look at the `pad` parameter).

Reply:
Well, I think here that would make the code less readable as I'd have to pass a pad element to partition(), and then in the loop ignore elements that are equal to the pad

Reply:
Well it's a good thing they provide a [`toolz.partition_all`](http://toolz.readthedocs.io/en/latest/api.html#toolz.itertoolz.partition_all) that I overlooked earlier.

```python
for neighbors in partition_all(max_neighbours, nodes):
    ...   # send message
```

Thoughts?

Reply:
Wouldn't it make more sense to stick to the builtin function that most devs will be familiar with already, and which will probably have a smaller chance of breaking in the future? I'm happy to change if you feel strong about it

Reply:
Your call.  I'm fine with whatever you choose.